server/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -717,6 +714,8 @@ public void onTimeout(TimeValue timeout) {
                 } catch (InterruptedException e) {
                     throw new ElasticsearchTimeoutException("Interrupted while waiting for initial discovery state");
                 }
+
+                NodeAndClusterIdStateListener.subscribeTo(observer);
```

Comment:
The name subscribeTo looks odd here, since this is a one time use. Perhaps the name should reflect that, like "getAndSetClusterIds"?

Reply:
Also, note that this is inside a block which will not run if the join has already happened. I think this needs to be moved outside this block.